### PR TITLE
feat: Complete MLX DeepFilterNet4 implementation with full PyTorch feature parity

### PR DESCRIPTION
**Pull request by sealad886**
# MLX DeepFilterNet4 Implementation + Workflow Fixes

## Summary

This PR completes the MLX (Apple Silicon native) implementation of DeepFilterNet4 with **full PyTorch feature parity**, plus fixes to GitHub Actions workflows and Python type errors. The MLX implementation is **1.5-1.9x faster** than PyTorch-MPS on Apple Silicon.

---

### GitHub Actions Fixes
- Replace `actions/checkout@v6` with `@v4` (v6 doesn't exist)
- Replace `actions/setup-python@v6` with `@v5` (v6 doesn't exist)
- Replace `actions/upload-artifact@v6` with `@v4` (v6 doesn't exist)
- Fix `build_demo.yml`: `ubuntu-25.0` → `ubuntu-latest` (25.0 doesn't exist)
- Fix `build_capi.yml`: `macos-12` → `macos-13/14` (12 is deprecated)
- Fix `build_wasm.yml`: remove invalid `matrix.target` reference

### Dependency Version Alignment
- Set workspace-level `ndarray = '=0.15.6'` to match hdf5-rust's requirement
- Update libDF, pyDF, pyDF-data Cargo.toml files for version consistency
- Pin pyo3 to 0.21, numpy to 0.21 for ABI compatibility

### Python Type Fixes (pyright 0 errors)
- Add `TYPE_CHECKING` guards and `cast()` for HDF5 Group typing
- Fix `init_df()` tuple unpacking (returns 4 values, not 3) in 5 files
- Fix `whisper_adapter.py`: get_tokenizer import and cast() for tensor operations
- Fix `train.py`: variable shadowing, GAN assertions, closure capture
- Fix `loss.py`: MLXWhisperBackend type narrowing with cast()
- Fix `io.py`: torchcodec metadata attribute access with getattr()
- Fix `utils.py`: `collections.Mapping` → `Mapping` for Python 3.10+
- Fix `sepm.py/stoi.py`: type annotations and boundary parameter ignores
- Fix WAcc scripts: pandas/librosa API compatibility

---

### Core Features
- ✅ **Lookahead configurations** for real-time variants (0, 1, 2 frame lookahead)
- ✅ **Post-filter integration** for musical noise reduction
- ✅ **Multi-resolution STFT loss** with PyTorch numerical parity
- ✅ **Complex gain output mode** for DfDecoder4
- ✅ **Model statistics tracking** (RunningStats, FeatureNormalizer, ModelStatistics)
- ✅ **Streaming/frame-by-frame inference** mode (StreamingDfNet4)

### Testing (243 tests passing)
- ✅ Comprehensive unit tests for all components
- ✅ End-to-end integration tests
- ✅ PyTorch numerical equivalence tests (tolerance: rtol=1e-4, atol=1e-5)

### Documentation
- ✅ README updated with MLX DeepFilterNet4 section
- ✅ 4 usage examples:
  - `examples/mlx_inference.py` - Single file enhancement
  - `examples/mlx_batch_enhance.py` - Batch processing
  - `examples/mlx_training.py` - Training loop demo
  - `examples/mlx_streaming.py` - Real-time streaming
- ✅ Performance benchmarks in `docs/BENCHMARKS.md`

---

## Performance Results (M3 Max)

| Batch Size | MLX Latency | PyTorch-MPS Latency | Speedup |
|------------|-------------|---------------------|---------|
| 1 | 15.2ms | 25.1ms | 1.65x |
| 4 | 23.8ms | 41.7ms | 1.75x |
| 8 | 31.4ms | 58.2ms | 1.85x |
| 16 | 47.1ms | 86.3ms | 1.83x |

---

## Test Results

```
243 passed in 45.23s
```

---

## Commits

1. `fix(mlx): resolve type errors in model.py`
2. `feat(mlx): add lookahead configurations for real-time variants`
3. `feat(mlx): integrate mask-based post-filter for musical noise reduction`
4. `feat(mlx): implement multi-resolution STFT loss with PyTorch parity`
5. `feat(mlx): add complex gain output mode to DfDecoder4`
6. `feat(mlx): add comprehensive integration tests`
7. `feat(mlx): add PyTorch numerical equivalence tests`
8. `feat(mlx): add streaming/frame-by-frame inference mode`
9. `feat(mlx): add model statistics tracking utilities`
10. `docs(mlx): add MLX usage examples and tutorials`
11. `docs(benchmarks): add MLX vs PyTorch performance documentation`

---

## Checklist

- [x] All tests pass (243/243)
- [x] Code follows project conventions (black, isort, flake8)
- [x] pyright reports 0 errors
- [x] Documentation updated
- [x] Examples provided
- [x] Performance benchmarked

### CODE REVIEW

File: DeepFilterNet/df/loss.py [hunk 1]
```diff
@@ -14,6 +14,9 @@
 from df.utils import angle, as_complex, get_device
 from libdf import DF
 
+if TYPE_CHECKING:
+    from df.whisper_adapter import MLXWhisperBackend
```
[the remark this answers is not in the document]
Comment (sealad886) [reply DeepFilterNet/df/loss.py]: This is a false positive. The import IS used by `cast('MLXWhisperBackend', ...)` calls at lines 475 and 555. The TYPE_CHECKING pattern with string literals in cast() is a common idiom to avoid circular imports while still providing type information. Added a noqa comment to suppress this false positive in commit e633fcd.

File: scripts/datasets/cleanup_hdf5_keys.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+"""
+
+import argparse
+import sys
+from typing import TYPE_CHECKING, cast
+
+import h5py
+
+if TYPE_CHECKING:
+    from h5py import Group
```
[the remark this answers is not in the document]
Comment (sealad886) [reply scripts/datasets/cleanup_hdf5_keys.py]: False positive - the import is used by `cast('Group', ...)` at line 77. Added noqa comment in e633fcd.

File: DeepFilterNet/df/loss.py [hunk 3]
```diff
@@ -14,6 +14,9 @@
...
+
```
[the remark this answers is not in the document]
Comment (sealad886) [reply DeepFilterNet/df/loss.py]: False positive - used by cast() at lines 475 and 555. Added noqa comment.

File: DeepFilterNet/df/scripts/split_hdf5.py [hunk 4]
```diff
@@ -3,11 +3,15 @@
...
 import shutil
 import sys
+from typing import TYPE_CHECKING, cast
 
 import h5py
 import numpy as np
 from icecream import ic
 
+if TYPE_CHECKING:
+    from h5py import Group
```
[the remark this answers is not in the document]
Comment (sealad886) [reply DeepFilterNet/df/scripts/split_hdf5.py]: False positive - used by cast() at line 40. Added noqa comment.
